### PR TITLE
Add support for sandbox and custom config names

### PR DESCRIPTION
**Pull request by samrocketman**
This adds two options to the Jenkins UI for configuring pipeline multibranch with defaults.

1. Customizing the config file name to something other than  `Jenkinsfile`.
2. Enabling groovy sandbox on the default Jenkinsfile being run.

Additionally, with these new options using the `@DataBoundSetter`, Job DSL is now available for configuring these settings.

Example Job DSL:

```groovy
multibranchPipelineJob('example') {
    factory {
        pipelineBranchDefaultsProjectFactory {
            // The ID of the default Jenkinsfile to use from the global Config File Management.
            scriptId('Jenkinsfile')

            // If enabled, the configured default Jenkinsfile will be run within a Groovy sandbox.
            useSandbox(true)
        }
    }
}
```

See also:

- [JENKINS-43285][JENKINS-43285] jobdsl for pipeline multibranch defaults
- [JENKINS-41108][JENKINS-41108] Posibility to specify the name of jenkinsfile

[JENKINS-41108]: https://issues.jenkins-ci.org/browse/JENKINS-41108
[JENKINS-43285]: https://issues.jenkins-ci.org/browse/JENKINS-43285

### CODE REVIEW

**Comment by samrocketman** [comment]
# Screenshot

screenshot of new plugin options

File: src/main/java/org/jenkinsci/plugins/pipeline/multibranch/defaults/DefaultsBinder.java [hunk 1]
```diff
@@ -50,6 +50,21 @@
...
 
+    private String scriptId;
+    private boolean useSandbox;
+
+    public DefaultsBinder(String scriptId, boolean useSandbox) {
+        this.scriptId = scriptId;
+        this.useSandbox = useSandbox;
+    }
+
+    public Object readResolve() {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/multibranch/defaults/DefaultsBinder.java]: Why is this here? Should never be used AFAICT.

File: src/main/java/org/jenkinsci/plugins/pipeline/multibranch/defaults/PipelineBranchDefaultsProjectFactory.java [hunk 2]
```diff
@@ -29,26 +29,82 @@
...
 
 /**
  * Recognizes and builds by default {@code Jenkinsfile}.
  */
-public class PipelineBranchDefaultsProjectFactory extends WorkflowBranchProjectFactory {
+public class PipelineBranchDefaultsProjectFactory extends AbstractWorkflowBranchProjectFactory {
     public static final String SCRIPT = "Jenkinsfile";
 
+    private String scriptId = SCRIPT;
+    private Boolean useSandbox = false;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/multibranch/defaults/PipelineBranchDefaultsProjectFactory.java]: Should be `boolean`. Otherwise you will get an NPE after load from previous settings.

**Comment by sgleske-ias** [comment]
@jglick I removed `readResolve()` from DefaultsBinder and PipelineBranchDefaultsProjectFactory classes.

**Comment by samrocketman** [comment]
I need to build and try out the latest code changes before this gets merged.

**Comment by samrocketman** [comment]
I have successfully tested the latest changes.

**Comment by samrocketman** [comment]
@vaimr do you still support this plugin or should I request ownership to maintain this plugin?

**Comment by yan796** [comment]
Very nice, this is exactly what I need, can we get a build to test it ?

**Comment by samrocketman** [comment]
@yan796 if you have maven and jdk8 installed, the build procedure is:

    mvn clean package

The plugin will be in `target/*.hpi`.

**Comment by yan796** [comment]
@samrockeyman thanks for your message. I am unfortunately only a end-user so I don't have anything setup to compile myself the plug-in.

I must be doing something wrong but when I generate a groovy script, the ID is randomly generated and read-only while the rest, i.e. name, and script, is editable.
I will check again....

**Comment by samrocketman** [comment]
You must customize the ID when creating the script.  Once it is created I don’t think you can change it.

**Comment by yan796** [comment]
@samrocketman you were write, it must be set before creating the script. Thanks a lot for your help.

**Comment by lasote** [comment]
I really like and want this feature ;)

**Comment by samrocketman** [comment]
@daniel-beck Please review this change.
